library/kotlin/src/io/envoyproxy/envoymobile/AsyncRequestFilter.kt

```diff
@@ -0,0 +1,36 @@
+package io.envoyproxy.envoymobile
+
+/*
+ * RequestFilter supporting asynchronous resumption.
+ */
+interface AsyncRequestFilter : RequestFilter {
+  /**
+   * Called by the filter manager once to initialize the filter callbacks that the filter should
+   * use.
+   *
+   * @param callbacks: The callbacks for this filter to use to interact with the chain.
+   */
+  fun setRequestFilterCallbacks(callbacks: RequestFilterCallbacks)
+
+  /**
+   * Invoked explicitly in response to an asynchronous resumeRequest() callback when filter
+   * iteration has been stopped. The parameters passed to this invocation will be a snapshot
+   * of any stream state that has not yet been forwarded along the filter chain.
+   *
+   * As with other filter invocations, this will be called on Envoy's main thread, and thus
+   * no additional synchronization is required between this and other invocations.
+   *
+   * @param headers: Headers, if `StopIteration` was returned from `onRequestHeaders`.
+   * @param data: Any data that has been buffered where `StopIterationAndBuffer` was returned.
+   * @param trailers: Trailers, if `StopIteration` was returned from `onRequestTrailers`.
+   * @param endStream: True, if the stream ended with the previous (and thus, last) invocation.
```

Comment:
Got it, so this will be true for trailers or `data, endStream: true` 👍

Reply:
Correct, in fact any of:
`onHeaders(... endStream: true)`
`onData(... endStream: true)`
`onTrailers(...)`